velox/functions/prestosql/aggregates/tests/MinMaxByAggregationTest.cpp

```diff
@@ -13,273 +13,988 @@
 #include "velox/exec/tests/utils/PlanBuilder.h"
 #include "velox/functions/prestosql/aggregates/AggregateNames.h"
 #include "velox/functions/prestosql/aggregates/tests/AggregationTestBase.h"
 
-using facebook::velox::exec::test::PlanBuilder;
+#include <fmt/format.h>
+
+using namespace facebook::velox::exec::test;
 
 namespace facebook::velox::aggregate::test {
 
 namespace {
 
-class MinMaxByAggregationTest : public AggregationTestBase {
+struct TestParam {
+  TypeKind valueType;
+  TypeKind comparisonType;
+  TypeKind groupByType = TypeKind::INVALID;
+};
+
+const std::unordered_set<TypeKind> kSupportedMinmaxByTypeSet = {
+    TypeKind::TINYINT,
+    TypeKind::SMALLINT,
+    TypeKind::INTEGER,
+    TypeKind::BIGINT,
+    TypeKind::REAL,
+    TypeKind::DOUBLE};
+
+#define EXECUTE_TEST_BY_GLOBAL_OR_GROUP(                                \
+    testFunc, valueType, comparisonType, isPartial)                     \
+  do {                                                                  \
+    switch (GetParam().groupByType) {                                   \
+      case TypeKind::TINYINT:                                           \
+        testFunc<valueType, comparisonType, int8_t>(isPartial);         \
+        break;                                                          \
+      case TypeKind::SMALLINT:                                          \
+        testFunc<valueType, comparisonType, int16_t>(isPartial);        \
+        break;                                                          \
+      case TypeKind::INTEGER:                                           \
+        testFunc<valueType, comparisonType, int32_t>(isPartial);        \
+        break;                                                          \
+      case TypeKind::BIGINT:                                            \
+        testFunc<valueType, comparisonType, int64_t>(isPartial);        \
+        break;                                                          \
+      case TypeKind::REAL:                                              \
+        testFunc<valueType, comparisonType, float>(isPartial);          \
+        break;                                                          \
+      case TypeKind::DOUBLE:                                            \
+        testFunc<valueType, comparisonType, double>(isPartial);         \
+        break;                                                          \
+      case TypeKind::INVALID:                                           \
+        testFunc<valueType, comparisonType, comparisonType>(isPartial); \
+        break;                                                          \
+      default:                                                          \
+        LOG(FATAL) << "Unsupported groupBy type of minmax_by(): "       \
+                   << mapTypeKindToName(GetParam().groupByType);        \
+    }                                                                   \
+  } while (0);
```

Comment:
What's the motivation for using a loop here?

Reply:
It is a common practice or patten in macro definition to avoid the potential compilation issue.